File: tournaments/nextPermutation/nextPermutation.js
const permutations = (arr) => {
  const collections = [];

  const collectPermutations = (str, prefix = '') => {
    if (!str.length) {
      return collections.push(prefix);
    }
    for (let i = 0; i < str.length; ++i) {
      if (~~str.slice(0,i).indexOf(str[i])) {
        collectPermutations(str.slice(0,i).concat(str.slice(i+1)), prefix+str[i]);
      }
    }
  };

  collectPermutations(arr);

  return collections;
};

const permutationsWithoutDuplicates = (arr) => {
  const collections = [];

  const shouldSwap = (str, start, curr) => {
    for (let i = start; i < curr; ++i) {
      if (str[i] === str[curr]) return false;
    }

    return true;
  };

  const findPermutations = (str, index = 0) => {
    const n = str.length;

    if (index >= n) {
      return collections.push(str);
    }

    for (let i = index; i < n; i++) {
      const check = shouldSwap(str, index, i);
      if (check) {
        const newStr = [...str];
        [newStr[index], newStr[i]] = [newStr[i], newStr[index]];
        findPermutations(newStr.join(''), index + 1);
      }
    }
  };

  findPermutations(arr);

  return collections;
};

File: tournaments/nextPermutation/nextPermutation.cpp
function nextPermutation(permutation) {

  for (var i = permutation.length - 2; i >= 0; i--) {
    if (permutation[i] < permutation[i + 1]) {
      var index = i + 1,
          t = permutation[i],
          suffix = [];
      for (var j = i + 1; j < permutation.length; j++) {
        if (permutation[j] > permutation[i] && permutation[j] < permutation[index]) {
          index = j;
        }
      }
      permutation[i] = permutation[index];
      permutation[index] = t;
      suffix = permutation.slice(i + 1);
      suffix.reverse();
      permutation = permutation.slice(0, i + 1).concat(suffix);
      return permutation;
    }
  }

  var compare = function(a, b) {
    return a - b;
  }

  permutation.sort(compare);
  return permutation;
}
